examples/gimli/src/gimli.rs

```diff
@@ -121,56 +121,77 @@ fn process_msg(message: &ByteSeq, mut s: State) -> (State, ByteSeq) {
-            msg_block_padded[block_len] = msg_block_padded[block_len] ^ U8(1u8);
-            s[11] = s[11] ^ U32(0x01000000u32); // s_2,3
-        }
-        s = absorb_block(msg_block_padded, s);
+        let msg_block = message.get_exact_chunk(rate, i);
+        let msg_block = Block::from_seq(&msg_block);
+        ciphertext = ciphertext.set_exact_chunk(rate, i, &(msg_block ^ key_block));
+        s = absorb_block(msg_block, s);
     }
 
+    // Handle final non-full block
+    let key_block = squeeze_block(s);
+    let last_block = message.get_remainder_chunk(rate);
+    let block_len = last_block.len();
+
+    // This pads the block if necessary
+    let msg_block_padded = Block::new();
+    let mut msg_block_padded = msg_block_padded.update_start(&last_block);
+
+    ciphertext = ciphertext.set_chunk(
+        rate,
+        num_chunks,
+        // the slice_range cuts off the last block if it is padded
+        &(msg_block_padded ^ key_block).slice_range(0..block_len),
+    );
+    msg_block_padded[block_len] = msg_block_padded[block_len] ^ U8(1u8);
+    s[11] = s[11] ^ U32(0x01000000u32); // s_2,3
+    s = absorb_block(msg_block_padded, s);
+
     (s, ciphertext)
 }
 
 fn process_ct(ciphertext: &ByteSeq, mut s: State) -> (State, ByteSeq) {
     let mut message = ByteSeq::new(ciphertext.len());
 
     let rate = Block::length();
-    let num_chunks = ciphertext.num_chunks(rate);
+    let num_chunks = ciphertext.num_exact_chunks(rate);
+
     for i in 0..num_chunks {
         let key_block = squeeze_block(s);
         let (block_len, ct_block) = ciphertext.get_chunk(rate, i);
 
         // This pads the ct_block if necessary.
         let ct_block_padded = Block::new();
         let ct_block_padded = ct_block_padded.update_start(&ct_block);
-        let msg_block = ct_block_padded ^ key_block;
+
         // Zero pad the block if necessary (replace bytes with zeros if block_len != rate).
-        let mut msg_block = Block::from_slice_range(&msg_block, 0..block_len);
+        let msg_block = Block::from_slice_range(&(ct_block_padded ^ key_block), 0..block_len);
 
         // Slice_range cuts off the msg_block to the actual length.
         message = message.set_chunk(rate, i, &msg_block.slice_range(0..block_len));
-        if i == num_chunks - 1 {
-            msg_block[block_len] = msg_block[block_len] ^ U8(1u8);
-            s[11] = s[11] ^ U32(0x01000000u32); // s_2,3
-        }
         s = absorb_block(msg_block, s);
     }
 
+    // Handle final non-full ct_block
+    let key_block = squeeze_block(s);
+    let ct_final = ciphertext.get_remainder_chunk(rate);
+    let block_len = ct_final.len();
+
+    let ct_block_padded = Block::new();
+    let ct_block_padded = ct_block_padded.update_start(&ct_final);
+    let msg_block = ct_block_padded ^ key_block;
+
+    let mut msg_block = Block::from_slice_range(&msg_block, 0..block_len);
+    message = message.set_chunk(rate, num_chunks, &msg_block.slice_range(0..block_len));
```

Comment:
This looks like you are slicing the `msg_block` to `block_len` twice. I think one should be enough.

Reply:
This has the same explanation as my above comment. I tried removing line 188 and adding a `mut` to line 186, but the program failed on all test cases, so that wasn't feasible.

Reply:
Ah well, it looks a little odd. But it's actually correct this way. The `set_chunk` on the message only needs to get the `block_len` part of the `msg_block` but we need the full, padded block later on. You could swap the two lines to make it look less odd. But both is correct.

Reply:
Ohk 👍